source/before-integrating/generate-a-key.html.md.erb

```diff
@@ -33,4 +33,13 @@ You have now generated your key pair, which will appear on your machine as 2 fil
 
 Once you have generated your key pair, you can [choose which user attributes your service can request][integrate.choose-user-attributes].
 
+## Publishing your public key(s) via a JWKS endpoint
+
+As an alternative to using a single fixed key pair, you may have multiple and publish the set of one or more public keys at an endpoint in the form of a [JWKS](https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets). This will enable you to do key rotations without having to directly contact us to make a configurtion update, avoiding potential downtime. In this case you must provide the URL that GOV.UK One Login is to call to `GET` the JWKS.
+
+Some considerations:
+
+* We may cache the return value from this endpoint for up to 24 hours. Therefore if you need to retract a public key due to a leak of your private key you should contact us for support and not simply remove the public key from your endpoint.
```

Comment:
There's some guidance on the rfc about the removal and presence of multiple keys on a jwks.json file.  We need to have the optional field 'kid' populated.

Reply:
```suggestion
* We will cache the public key on key ID for 24 hours. Therefore if you need to retract a public key you should contact us for support and not simply remove the public key from your endpoint.
```
Just to make it more specific as caching is built into the implementation

Reply:
@billrdunn, I agree the line about their private key is probably unnecessarily specific. I think it's important that the RP knows in general it's fine to remove a key from the endpoint though, only that it won't take immediate effect. We're going to have to rethink that section a little once we have clarity on how to communicate how the RP should raise the support issue (right now the process is only documented in a welcome pack apparently, not the tech docs).

Is there no scenario where we might find ourselves refreshing the cache sooner e.g. for support or maintenance reasons? On a cache miss, do we not update all keys returned by the endpoint? In my mind it makes sense to make the weakest commitment that gets across they can't rely that an edit to an existing key will be reflected until 24 hours from when they make the change.
```suggestion
* We will cache keys returned from this endpoint on `kid` (Key ID) for up to 24 hours. Therefore if you need to immediately retract a public key you should contact us for support and not simply remove the public key from your endpoint.
```

If we fail validation with cached key, do we try the endpoint again. If not, it might make sense to also tell the RP they either need to contact support if they need to alter a key for a given _kid_ or use a new _kid_.

Reply:
We only fetch / cache a key with `kid` equal to what is given in the header of the signed JWT. So if validation fails with a cached key, yet the `kid`s match, that means a crypto issue so we do not re-fetch. 

All we require is that the RP provides the `kid` in the header, and the corresponding JWKS with that `kid` is available on their endpoint. I don't think we need to tell them to contact support, it's their responsibility to ensure that their JWKS endpoint is up to date. Although we could warn them that if they don't keep the key up to date, their service will be unreachable.

Your suggested change looks good to me 👍

Reply:
Also if the RP needs to retract a key, they can just add a new one to the endpoint and start using it straight away. Once the new one is in use they can remove the old one. No need to contact us I don't think.